pkg/ingester/activeseries/model/matchers.go

```diff
@@ -3,28 +3,147 @@
+
+	// matchersForExactValue is a list of matchers that have at least one SetMatcher or exact value matcher.
+	matchersForExactValue := make([]indexedMatchers, 0, len(matchersConfig.config))
+	// As we scan them, keep track of the amount of different values matched per label name.
+	valuesMatchedPerLabelName := map[string]map[string]struct{}{}
+	for name, ms := range matchersConfig.config {
+		index := len(names)
+		names = append(names, name)
+
+		if len(ms) == 1 && !ms[0].Matches("") {
+			// This is a single matcher that matches on a single label.
+			// We can optimize it by treating it as a single label matcher.
+			singleLabelMatchers[ms[0].Name] = append(singleLabelMatchers[ms[0].Name], indexedMatchers{labelsMatchers: ms, index: index})
+			continue
+		}
+
+		optimized := false
+		for _, m := range ms {
+			if m.Type == labels.MatchEqual && m.Value != "" {
+				matchersForExactValue = append(matchersForExactValue, indexedMatchers{labelsMatchers: ms, index: index})
+				if valuesMatchedPerLabelName[m.Name] == nil {
+					valuesMatchedPerLabelName[m.Name] = make(map[string]struct{})
+				}
+				valuesMatchedPerLabelName[m.Name][m.Value] = struct{}{}
+				optimized = true
+				break
+			} else if sm := m.SetMatches(); len(sm) > 0 {
+				matchersForExactValue = append(matchersForExactValue, indexedMatchers{labelsMatchers: ms, index: index})
+				if valuesMatchedPerLabelName[m.Name] == nil {
+					valuesMatchedPerLabelName[m.Name] = make(map[string]struct{})
+				}
+				for _, v := range sm {
+					valuesMatchedPerLabelName[m.Name][v] = struct{}{}
+				}
+				optimized = true
+				break
+			}
+		}
+		if !optimized {
+			unoptimized = append(unoptimized, indexedMatchers{labelsMatchers: ms, index: index})
+		}
+	}
+
+	// Get label names and sort them by popularity (desc).
+	// We select labels that have the most different values matched first to make the decision tree higher and lower.
+	labelNames := slices.Collect(maps.Keys(valuesMatchedPerLabelName))
+	slices.SortFunc(labelNames, func(a, b string) int {
+		return cmp.Compare(len(valuesMatchedPerLabelName[b]), len(valuesMatchedPerLabelName[a]))
+	})
+
+	// valueMatchers is a map of label names to a map of values to indexedMatchers.
+	valueMatchers := make(map[string]map[string][]indexedMatchers, len(labelNames))
+	// We can probably optimize this 3 nested loops, but:
+	// - This is not the hot path
+	// - Probably there are not so many labelNames.
+	// - Probably there are not so many labelsMatchers
+	for _, name := range labelNames {
+		for mi := 0; mi < len(matchersForExactValue); {
+			ms := matchersForExactValue[mi]
+			removed := false
+			for lmi, m := range ms.labelsMatchers {
+				if m.Name != name {
+					continue
+				}
+				var sm []string
+				if m.Type == labels.MatchEqual && m.Value != "" {
+					// This is not a set matcher, but a simple equality matcher.
+					// We'll treat it as a set matcher with a single value.
+					stack := [1]string{m.Value}
+					sm = stack[:]
```

Comment:
I don't think there's an actual difference between this and just `sm := []string{m.Value}`. In both cases escape analysis is going to determine whether the underlying array should be allocated in the stack or escapes to the heap.

Reply:
Yep, you're right, done in https://github.com/grafana/mimir/pull/12184/commits/fbfc94c6bd276dfb6ce65ec6dc933ecbfccd4759
```
pkg/ingester/activeseries/model/matchers.go:93:19: []string{...} does not escape
```